uniffi_udl/src/finder.rs

```diff
@@ -129,43 +129,89 @@ impl TypeFinder for weedle::TypedefDefinition<'_> {
                 },
             )
         } else {
+            let typedef_type = match &self.type_.type_ {
+                weedle::types::Type::Single(weedle::types::SingleType::NonAny(
+                    weedle::types::NonAnyType::Identifier(weedle::types::MayBeNull {
+                        type_: i,
+                        ..
+                    }),
+                )) => i.0,
+                _ => bail!("Failed to get typedef type: {:?}", self),
+            };
+
             let module_path = types.module_path();
             let name = self.identifier.0.to_string();
-            let ty = match attrs.rust_kind() {
-                Some(RustKind::Object) => Type::Object {
-                    module_path,
-                    name,
-                    imp: ObjectImpl::Struct,
-                },
-                Some(RustKind::Trait) => Type::Object {
-                    module_path,
-                    name,
-                    imp: ObjectImpl::Trait,
-                },
-                Some(RustKind::CallbackTrait) => Type::Object {
-                    module_path,
-                    name,
-                    imp: ObjectImpl::CallbackTrait,
-                },
-                Some(RustKind::Record) => Type::Record { module_path, name },
-                Some(RustKind::Enum) => Type::Enum { module_path, name },
-                Some(RustKind::CallbackInterface) => Type::CallbackInterface { module_path, name },
-                // must be external
+
+            let ty = match attrs.external_tagged() {
                 None => {
-                    let kind = attrs.external_kind().expect("External missing kind");
-                    let tagged = attrs.external_tagged().expect("External missing tagged");
-                    Type::External {
-                        name,
-                        namespace: "".to_string(), // we don't know this yet
-                        module_path: attrs.get_crate_name(),
-                        kind,
-                        tagged,
+                    // Not external, not custom, not Rust - so we basically
+                    // pretend it is Rust, thus soft-deprecating it.
+                    // We use `type_`
+                    match typedef_type {
+                        "dictionary" | "record" | "struct" => Type::Record {
+                            module_path,
+                            name,
+                        },
+                        "enum" => Type::Enum {
+                            module_path,
+                            name,
+                        },
+                        "custom" => panic!("don't know builtin"),
+                        "interface" | "impl" => Type::Object {
```

Comment:
I should call out that I added a couple of names compared to the Rust attribute. In particular, `struct` as an alias for `record` above, and `impl` as an alias for `interface` here. My thinking was that these names more closely match what a Rust programmer is likely to think of - ie, a `derive(uniffi::Record)` lives on a `struct` defn while `uniffi::export` lives on an `impl` block.

On the negative side, `impl` is also used for traits.

I personally quite like these names, but don't feel too strongly about it, so will happily concede to someone who strongly dislikes them!